### PR TITLE
Rewritten TeamSpec::balance_extents() to balance more than two dimensions

### PR DESCRIPTION
**Pull request by ddiefenthaler**
I've rewritten balance_extents to use a different algorithm.

### CODE REVIEW

**Comment by ddiefenthaler** [comment]
The idea behind the algorithm is to equally distribute the prime factors across the dimensions as good as possible. Starting with the largest factors they get multiplied with `this->extents[i]` with `i` growing.
As long as there are enough remaining prime factors this condition should be true:
`this->_extents[0] >= this->_extents[1] >= ... >= this->_extents[MaxDimensions-1]`
If the last iteration over `this->_extents` is incomplete the direction gets reversed to even the effect out. (Starting with `this->_extents[MaxDimensions-1]` downwards)

Thinking about it, it may make sense to change the direction each iteration over `this->_extents`,
but I'm not sure about this... I can't see what would happen after the second iteration.

**Review by fuchsto — CHANGES_REQUESTED** [review]
Thank you, but please implement additional unit tests to cover the new functionality.

**Comment by fuchsto** [comment]
Addresses #228

**Comment by ddiefenthaler** [comment]
I added a unit test,
but I think that the algorithm should still be changed once more.
I assume that my approach only works well if the number of units factorizes in not to many different factors.
Apart from that the algorithm does not consider any constraints.
@fuchsto It would be nice if you could clarify how balance_extents should behave.

**Comment by ddiefenthaler** [comment]
I added another commit, I think this is in a mergable state now.

I am sure there are more optimizations possible, but for now this works too.

I removed the adjustment of `_is_linear`, that led to `MatrixTest.CopyRow` failing,
but I think this is another error so I will open a seperate issue for that.

**Comment by ddiefenthaler** [comment]
I think this Pull request is fine now,
@fuchsto can you review this again?

File: dash/include/dash/TeamSpec.h [hunk 1]
```diff
@@ -218,39 +219,56 @@ class TeamSpec :
...
-          this->_extents[0] = extent_x;
-          this->_extents[1] = extent_y;
-        }
+    // Equally distribute factors to extents.
+    // Start with the largest factors and multiply them onto the lowest value
+    for (auto it = teamsize_prime_factors.rbegin(); it != teamsize_prime_factors.rend(); ++it) {
+      DASH_LOG_TRACE("TeamSpec.balance_extents()",
+                     "factor:", it->first, "x", it->second);
+      for (auto i = 1; i < it->second + 1; ++i) {
+        new_extents.insert(it->first * *new_extents.begin());
```
Comment (fmoessbauer) [inline dash/include/dash/TeamSpec.h:259]: You are not a fan of brackets, are you? ;)

**Review by fuchsto — APPROVED** [review]

**Comment by fuchsto** [comment]
@ddiefenthaler Looking good!

**Comment by fuchsto** [comment]
@ddiefenthaler I don't care too much about efficiency of this implementation for now. Integer factorization is NP-hard, anyways.
This function is supposed to be called once for a reasonably large unit of work, as in domain decomposition for a 2048x2048 matrix-matrix multiplication. Even if it was inefficient, it would not notably affect the overall time to completion.

**Pull request merged**